cloudnetwork/v1/types.go

```diff
@@ -0,0 +1,75 @@
+package v1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// CloudPrivateIPConfig performs assignments of a private IP address to the
+// primary NIC associated with cloud VMs. This is done by specifying the IP and
+// Kubernetes node which the IP should be assigned to. This CRD is intended to
+// be used by the network plugin which manages the cluster network. The spec
+// side represents the desired state requested by the network plugin, and the
+// status side represents the current state that this CRD's controller has
+// executed. No users will have permission to modify it, and if a cluster-admin
+// decides to edit it for some reason, their changes will be overwritten the
+// next time the network plugin reconciles the object. Note: the CR's name
+// must specify the requested private IP address (can be IPv4 or IPv6).
+// The IP format will be validated in an admission controller.
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +k8s:openapi-gen=true +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:path=cloudprivateipconfig,scope=Cluster
+type CloudPrivateIPConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// spec is the definition of the desired private IP request.
+	// +kubebuilder:validation:Required
+	// +required
+	Spec CloudPrivateIPConfigSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+	// status is the observed status of the desired private IP request. Read-only.
+	// +kubebuilder:validation:Optional
+	// +optional
+	Status CloudPrivateIPConfigStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// CloudPrivateIPConfigStatus specifies the requested private IP, its corresponding node assignment,
+// together with its assignment condition.
+// +k8s:openapi-gen=true
+type CloudPrivateIPConfigStatus struct {
```

Comment:
Remind me.  Can you scan the nodes to find which nodes currently have IP?  It's ok to maintain a status of which node actually has the IP here to make the controller slightly easier to write.

Reply:
Shouldn't the informer's update func handle that smoothly? I would only need to compare `old.spec.node` vs `new.spec.node`, which is why I removed the node from the status field.

Reply:
> Shouldn't the informer's update func handle that smoothly? I would only need to compare `old.spec.node` vs `new.spec.node`, which is why I removed the node from the status field.

You need to consider cases of uncontrolled controller restart.